core/playlistmanager.go

```diff
@@ -0,0 +1,262 @@
+	manifestID     ManifestID
+	// Live playlist used for broadcasting
+	masterLivePList *m3u8.MasterPlaylist
+	// Full playlist is stored into OS for later reference
+	masterFullPList *m3u8.MasterPlaylist
+	mediaLiveLists  map[StreamID]*m3u8.MediaPlaylist
+	mediaFullLists  map[StreamID]*m3u8.MediaPlaylist
+	mapSync         *sync.Mutex
+}
+
+// NewBasicPlaylistManager create new BasicPlaylistManager struct
+func NewBasicPlaylistManager(manifestID ManifestID,
+	storageSession drivers.OSSession) *BasicPlaylistManager {
+
+	bplm := &BasicPlaylistManager{
+		storageSession:  storageSession,
+		manifestID:      manifestID,
+		masterLivePList: m3u8.NewMasterPlaylist(),
+		masterFullPList: m3u8.NewMasterPlaylist(),
+		mediaLiveLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mediaFullLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mapSync:         &sync.Mutex{},
+	}
+	return bplm
+}
+
+func (mgr *BasicPlaylistManager) ManifestID() ManifestID {
+	return mgr.manifestID
+}
+
+func (mgr *BasicPlaylistManager) GetSession() drivers.OSSession {
+	return mgr.storageSession
+}
+
+// InsertHLSSegment ...
+func (mgr *BasicPlaylistManager) InsertHLSSegment(streamID StreamID,
+	seg *stream.HLSSegment) (*net.TypedURI, error) {
+
+	if err := mgr.isRightStream(streamID); err != nil {
+		return nil, err
+	}
+	sid := string(streamID)
+	fmpl := mgr.getFullPL(streamID)
+	if fmpl == nil {
+		panic("no full media pl")
+	}
+	if mgr.hasSegment(fmpl, seg.SeqNo) {
+		return nil, fmt.Errorf("Already saved")
+	}
+	turi, _, err := mgr.storageSession.SaveData(sid, seg.Name, seg.Data)
+	if err != nil {
+		glog.Error("Error saving segment", err)
+	} else {
+		turi.SeqNo = seg.SeqNo
+		mgr.insertLink(turi, seg.Duration)
+	}
+	return turi, err
+}
+
+func (mgr *BasicPlaylistManager) isRightStream(streamID StreamID) error {
+	mid, err := streamID.ManifestIDFromStreamID()
+	if err != nil {
+		return err
+	}
+	if mid != mgr.manifestID {
+		return fmt.Errorf("Wrong sream id (%s), should be %s", streamID, mid)
+	}
+	return nil
+}
+
+func (mgr *BasicPlaylistManager) hasSegment(mpl *m3u8.MediaPlaylist, seqNum uint64) bool {
+	for _, seg := range mpl.Segments {
+		if seg != nil && seg.SeqId == seqNum {
+			return true
+		}
+	}
+	return false
+}
+
+func (mgr *BasicPlaylistManager) getLivePL(streamID StreamID) *m3u8.MediaPlaylist {
+	mgr.mapSync.Lock()
+	defer mgr.mapSync.Unlock()
+	mpl := mgr.mediaLiveLists[streamID]
+	var err error
+	if mpl == nil {
+		mpl, err = m3u8.NewMediaPlaylist(LIVE_LIST_LENGTH, LIVE_LIST_LENGTH)
+		if err != nil {
+			glog.Error(err)
+		} else {
+			mgr.mediaLiveLists[streamID] = mpl
+			sid := string(streamID)
+			vParams := ffmpeg.VideoProfileToVariantParams(ffmpeg.VideoProfileLookup[streamID.GetRendition()])
+			url, err := mgr.savePlayList(mpl, streamID, sid+".m3u8")
+			if err == nil {
+				mgr.masterLivePList.Append(url, mpl, vParams)
+				mid, _ := streamID.ManifestIDFromStreamID()
+				_, _, err := mgr.storageSession.SaveData(sid, string(mid)+".m3u8",
```

Comment:
it's not just get operation, it's `get or create and get`, it creates new media playlist if it not exists, adds it to master playlist - so master pl modified, and we need to save it

Reply:
why? it is 'return existing or create new one' operation specifically

Reply:
Read operations shouldn't mutate. We can do an "insert or create-then-insert" when we put a segment into a (possibly empty) playlist, but the act of attempting to retrieve an empty playlist should not create a persistent playlist in OS.

Reply:
Will it help if I'll rename this function to `createOrGetPL`? It is purpose of this function to create new playlist, or get one if it is already exists.

Reply:
Why don't we create the playlist as needed when we attempt to insert a segment into a nonexistent playlist? That seems much less error prone, and easier to reason about.

I'm still not sure if we really should be writing the manifest to OS after *each* segment. Especially for remote/S3 calls, that could be quite slow and it's done in the middle of the transcode path, eg InsertLink. Likewise, doing it asynchronously is racy.

Reply:
> Why don't we create the playlist as needed when we attempt to insert a segment into a nonexistent playlist? That seems much less error prone, and easier to reason about.

This is exactly what we're doing
You're right, we should move save operation to own goroutine. And we can save manifest every 10 segments, with risk that manifest will be missing up to 10 last segments in case something goes wrong

Reply:
> This is exactly what we're doing

No, it's being created during the *read* operation. A more explicit approach, following your current code, would be:

```
func (pm *BasicPlaylistManager) addToMediaPlaylist(url, streamId, ...) {
  pl = pm.getMediaPlaylist(streamId)
  if pl == nil {
    pl = pm.addMediaPlaylist(streamId)
 }
 seg = ...
 pl.appendSegment(seg)
}
```

Also I wonder if the `live` parameter could be a property of the media playlist itself so it doesn't need to be passed at every invocation, and can be automatically trimmed when a new segment is appended.